instruqt/delivering-as-an-appliance/03-collecting-configuration/assignment.md

```diff
@@ -11,16 +11,355 @@ notes:
+
+There are no sensible default values for the options in the "SlackerNews Core"
+group, so let's move on temporarily to the "Certificates" group. The
+application will always run over TLS, but we can default to generating a
+self-signed certificate.
+
+```yaml
+        - name: certificate_source
+          type: select_one
+          title: Certificate Source
+          default: generate_internal
+          items:
+            - name: generate_internal
+              title: Generate
+            - name: upload_existing
+              title: Upload
+```
+
+Add those changes as well. We'll see a few default values in the final
+configuration.
+
+### Displaying Fields Conditionally
+
+If we accept the default self-signed certificate option, we don't need to
+upload the three files that support using our own. We could use the group
+description to explain this. A better choice for our users, though, would be
+to show those fields only if they're needed. We can do this by adding a
+property called `when` to those fields, and using a bit of templating to
+determine when the fields should be shown.
+
+```yaml
+        - name: tls_cert
+          title: Certificate
+          type: file
+          when: '{{repl ConfigOptionEquals "certificate_source" "upload_existing"}}'
+        - name: tls_key
+          title: Private Key
+          type: file
+          when: '{{repl ConfigOptionEquals "certificate_source" "upload_existing"}}'
+        - name: tls_ca
+          title: Signing Authority
+          type: file
+          when: '{{repl ConfigOptionEquals "certificate_source" "upload_existing"}}'
+```
+
+The templating under `when` is based on Go templates. In this case, we're
+looking at the value of the `certificate_source` option and only collecting
+the files if the user wants to upload their own certificate. In the next
+section of the lab we'll look at templating a bit more.
+
+Make these changes to the three conditional fields in the "Certificates"
+group. Like the other features we're looking at, you'll see a few more
+conditional fields in the final configuration.
+
+### Validating Input
+
+The last feature we'll look at is input validation. Validation makes sure the
+user sets a value in the correct format (unfortunately, we can't make sure
+it's the _right_ value, the customer has to handle that). Fields validation
+uses regular expressions, though the `Config` definition provides for
+Replicated adding more options in the future.
+
+Let's return the "Ingress Hostname" field. It's important for this field to be
+a valid hostname. Though we can't check that it's a real domain here (that's
+the type of thing preflight checks are for), we can make sure it's a valid
+name.
+
+```yaml
+          validation:
+            regex:
+              pattern: ^(([a-zA-Z0-9]([a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)\.)+([a-zA-Z]{2,6}|[a-zA-Z0-9-]{2,30}\.[a-zA-Z]{2,3})$
+              message: Please enter a valid hostname
+
+```
```

Comment:
^ could be helpful to include the full ingress hostname item in this YAML sample so it's easier to visually confirm where in the file to paste this new yaml

Reply:
I want them to be able to easily copy/paste…so I just added some text pointing them in the right direction.